lib/lita/handlers/confirmation.rb

```diff
@@ -1,24 +1,130 @@
+require "net/smtp"
+require "rotp"
+require "rqrcode"
+
 module Lita
   module Handlers
     class Confirmation < Handler
+      # block - never use 2fa; allow; require - only allow 2fa
+      config :twofactor_default, required: false, type: [:block, :allow, :require], default: :block
+
+      # If users can re-enroll or remove themselves, there is no true "second factor"
+      # protection as control of Slack account lets an attacker re-enroll, obtain a new TOTP secret,
+      # and generate OTP passwords. This can be set to false during a transition to 2FA.
+      config :twofactor_secure, required: false, type: [TrueClass, FalseClass], default: true
+
+      config :smtp_host, required: false, type: String, default: 'localhost'
+      config :smtp_port, required: false, type: Fixnum, default: 25
+      config :from_email, required: false, type: String, default: 'example@example.com'
+
       route /^confirm\s+([a-f0-9]{6})$/i, :confirm, command: true, help: {
-        t("help.key") => t("help.value")
+        t("confirm_help.key") => t("confirm_help.value")
       }
 
+      route /^confirm\s+([a-f0-9]{6})\s+([0-9]{6})$/i, :totp_confirm, command: true, help: {
+        t("confirm_totp_help.key") => t("confirm_totp_help.value")
+      }
+
+      route /^confirm\s+2fa\s+enroll\s+(.*)\s*$/i,
+        :enroll, command:true, help: { t("enroll_help.key") => t("enroll_help.value") }
+
+      route /^confirm\s+2fa\s+remove$/i, :remove_self, command: true, help: {
+        t("remove_help.key") => t("remove_help.value")
+      }
+
+      route /^confirm\s+2fa\s+remove\s+(\S+)$/i, :remove, command: true
+
       def confirm(response)
         code = response.matches[0][0]
+        command = Extensions::Confirmation::UnconfirmedCommand.find(code)
+
+        unless command
```

Comment:
This (and others like it) can be shortened to: `return response.reply(t("invalid_code", code: code)) unless command`